FWCore/Framework/src/ProductResolversFactory.cc

```diff
@@ -0,0 +1,253 @@
+      return std::make_shared<AliasProductResolver>(
+          std::move(bd), dynamic_cast<DataManagingOrAliasProductResolver&>(*iResolvers[index]));
+    }
+    std::shared_ptr<ProductResolverBase> makeSwitchProducerProduct(
+        std::shared_ptr<BranchDescription const> bd,
+        ProductRegistry const& iReg,
+        std::vector<std::shared_ptr<ProductResolverBase>> iResolvers) {
+      ProductResolverIndex index = iReg.indexFrom(bd->switchAliasForBranchID());
+      assert(index != ProductResolverIndexInvalid);
+
+      return std::make_shared<SwitchProducerProductResolver>(
+          std::move(bd), dynamic_cast<DataManagingOrAliasProductResolver&>(*iResolvers[index]));
+    }
+
+    std::shared_ptr<ProductResolverBase> makeSwitchAliasProduct(
+        std::shared_ptr<BranchDescription const> bd,
+        ProductRegistry const& iReg,
+        std::vector<std::shared_ptr<ProductResolverBase>> iResolvers) {
+      ProductResolverIndex index = iReg.indexFrom(bd->switchAliasForBranchID());
+      assert(index != ProductResolverIndexInvalid);
+
+      return std::make_shared<SwitchAliasProductResolver>(
+          std::move(bd), dynamic_cast<DataManagingOrAliasProductResolver&>(*iResolvers[index]));
+    }
+
+    std::shared_ptr<ProductResolverBase> makeParentProcessProduct(std::shared_ptr<BranchDescription const> bd) {
+      return std::make_shared<ParentProcessProductResolver>(std::move(bd));
+    }
+
+    void addProductOrThrow(std::shared_ptr<ProductResolverBase> iResolver,
+                           std::vector<std::shared_ptr<ProductResolverBase>>& oResolvers,
+                           ProductRegistry const& iReg) {
+      assert(bool(iResolver));
+      BranchDescription const& bd = iResolver->branchDescription();
+      assert(!bd.className().empty());
+      assert(!bd.friendlyClassName().empty());
+      assert(!bd.moduleLabel().empty());
+      assert(!bd.processName().empty());
+      auto index = iReg.indexFrom(bd.branchID());
+      if (oResolvers[index]) {
+        throw Exception(errors::InsertFailure, "AlreadyPresent")
+            << "Problem found while adding a new ProductResolver, "
+            << "product already exists for (" << bd.friendlyClassName() << "," << bd.moduleLabel() << ","
+            << bd.productInstanceName() << "," << bd.processName() << ")\n";
+      }
+      oResolvers[index] = std::move(iResolver);
+    }
+
+    std::shared_ptr<ProductResolverBase> makeForPrimary(BranchDescription const& bd,
+                                                        ProductRegistry const& iReg,
+                                                        ProductResolverIndexHelper const& iHelper) {
+      auto cbd = std::make_shared<BranchDescription const>(bd);
+      if (bd.produced()) {
+        using namespace std::literals;
+        if (bd.moduleLabel() == "source"sv) {
+          return makeSourceProduct(cbd);
+        } else if (bd.onDemand()) {
+          if (bd.isTransform()) {
```

Comment:
Do we not need to assert `branchType_ == InEvent` here anymore?

Reply:
added